lib/services/expiry/expiry.go

```diff
@@ -0,0 +1,237 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package expiry
+
+import (
+	"context"
+	"log/slog"
+	"time"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/auth/authclient"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils/interval"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+)
+
+var (
+	// scanInterval is the interval at which the expiry checker scans for access requests.
+	scanInterval = time.Minute * 5
+
+	// pendingRequestGracePeriod is the grace period used when checking a pending request's expiry
+	// as the expiry time may be extended on approval.
+	pendingRequestGracePeriod = time.Second * 40
+)
+
+const (
+	semaphoreName       = "auth.expiry"
+	semaphoreExpiration = time.Minute * 5
+	semaphoreJitter     = time.Minute
+
+	// minPageDelay is the minimum delay between processing each page of access requests.
+	minPageDelay           = time.Millisecond * 200
+	accessRequestPageLimit = 100
+	// maxExpiresPerCycle is an arbitrary limit on the number of requests to expire per cycle
+	// to prevent any one auth server holding the lease for more than a couple of minutes.
+	maxExpiresPerCycle = 120
+)
+
+// Config provides configuration for the expiry server.
+type Config struct {
+	// Log is the logger.
+	Log *slog.Logger
+	// Emitter is an events emitter, used to submit discrete events.
+	Emitter apievents.Emitter
+	// AccessPoint is a expiry access point.
+	AccessPoint authclient.ExpiryAccessPoint
+	// Clock is the clock used to calculate expiry.
+	Clock clockwork.Clock
+	// HostID is a unique ID of this host.
+	HostID string
+}
+
+// CheckAndSetDefaults checks required fields and sets default values.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Emitter == nil {
+		return trace.BadParameter("no Emitter configured for expiry")
+	}
+	if c.AccessPoint == nil {
+		return trace.BadParameter("no AccessPoint configured for expiry")
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	return nil
+}
+
+// Service is a expiry service.
+type Service struct {
+	*Config
+}
+
+// New initializes a expiry service
+func New(cfg *Config) (*Service, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s := &Service{
+		Config: cfg,
+	}
+	return s, nil
+}
+
+// Run starts the expiry service.
+func (s *Service) Run(ctx context.Context) error {
+	semCfg := services.SemaphoreLockConfigWithRetry{
```

Comment:
Nit: In cases like this, to keep the function smaller I would consider having a helper method on `Service` that returns the semaphore config, this would make this `Run()` method slightly cleaner cause right now basically half of it is this giant object.

```go
func (s *Service) getSemaphoreCfg() services.SemaphoreLockConfigWithRetry {
    return services.SemaphoreLockConfigWithRetry{
    ...
    }
}
```